omnigibson/object_states/folded.py

```diff
@@ -1,50 +1,150 @@
 import numpy as np
+from collections import namedtuple
+from scipy.spatial import ConvexHull, distance_matrix
 
 from omnigibson.macros import create_module_macros
 from omnigibson.object_states.object_state_base import BooleanState, AbsoluteObjectState
-from omnigibson.utils.object_state_utils import (
-    calculate_projection_area_and_diagonal,
-    calculate_projection_area_and_diagonal_maximum,
-    calculate_smoothness,
-)
 
 # Create settings for this module
 m = create_module_macros(module_path=__file__)
 
 # Criterion #1: the threshold on the area ratio of the convex hull of the projection on the XY plane
-m.AREA_THRESHOLD = 0.75
+m.FOLDED_AREA_THRESHOLD = 0.75
+m.UNFOLDED_AREA_THRESHOLD = 0.9
 
 # Criterion #2: the threshold on the diagonal ratio of the convex hull of the projection on the XY plane
-m.DIAGONAL_THRESHOLD = 0.9
+m.FOLDED_DIAGONAL_THRESHOLD = 0.85
+m.UNFOLDED_DIAGONAL_THRESHOLD = 0.9
 
-# Criterion #3: the percentage of face normals that need to be close to the z-axis.
+# Criterion #3: the percentage of face normals that need to be close to the z-axis
 m.NORMAL_Z_PERCENTAGE = 0.5
 
-class Folded(AbsoluteObjectState, BooleanState):
-
+# Whether to visualize the convex hull of the projection on the XY plane
+m.DEBUG_CLOTH_PROJ_VIS = False
+
+# Angle threshold for checking smoothness of the cloth; surface normals need to be close enough to the z-axis
+m.NORMAL_Z_ANGLE_DIFF = np.deg2rad(45.0)
+
+"""
+FoldedLevelData contains the following fields:
+    smoothness (float): percentage of surface normals that are sufficiently close to the z-axis
+    area (float): the area of the convex hull of the projected points compared to the initial unfolded state
+    diagonal (float): the diagonal of the convex hull of the projected points compared to the initial unfolded state
+"""
+FoldedLevelData = namedtuple("FoldedLevelData", ("smoothness", "area", "diagonal"))
```

Comment:
Oh nice!! maybe we should do the same for ContainedParticles state (and in general, any intermediate state)?